src/main/java/ch/njol/skript/expressions/ExprBeaconValues.java

```diff
@@ -0,0 +1,248 @@
+				return false;
+			}
+			setExpr(new EventValueExpression<>(Block.class));
+		}
+		if (valueType == BeaconValues.PRIMARY || valueType == BeaconValues.SECONDARY)
+			isEffect = true;
+
+		return true;
+	}
+
+	@Override
+	protected Object @Nullable [] get(Event event, Block[] source) {
+		return get(source, new Getter<Object, Block>() {
+			@Override
+			public @Nullable Object get(Block block) {
+				Beacon beacon  = (Beacon) block.getState();
+				switch (valueType) {
+					case PRIMARY -> {
+						if (PAPER_EVENTS && event instanceof PlayerChangeBeaconEffectEvent changeEvent) {
+							return changeEvent.getPrimary();
+						} else if (beacon.getPrimaryEffect() != null) {
+							return beacon.getPrimaryEffect().getType();
+						}
+					}
+					case SECONDARY-> {
+						if (PAPER_EVENTS && event instanceof PlayerChangeBeaconEffectEvent changeEvent) {
+							return changeEvent.getSecondary();
+						} else if (beacon.getSecondaryEffect() != null) {
+							return beacon.getSecondaryEffect().getType();
+						}
+					}
+					case RANGE -> {return beacon.getEffectRange();}
+					case TIER -> {return beacon.getTier();}
+				};
+				return null;
+			}
+		});
+	}
+
+	@Override
+	public Class<?> @Nullable [] acceptChange(ChangeMode mode) {
+		if (mode == ChangeMode.REMOVE_ALL || valueType == BeaconValues.TIER) {
+			return null;
+		}
+		if (isEffect && (mode == ChangeMode.ADD || mode == ChangeMode.REMOVE)) {
+			return null;
+		}
+		return CollectionUtils.array(Object.class);
+	}
+
+	@Override
+	public void change(Event event, Object @Nullable [] delta, ChangeMode mode) {
+		PotionEffectType providedEffect = null;
+		double providedRange = 0;
+		if (delta != null && delta[0] != null) {
+			if (isEffect && delta[0] instanceof PotionEffectType potionEffectType) {
+				providedEffect = potionEffectType;
+			} else if (valueType == BeaconValues.RANGE) {
+				providedRange = (double) ((long) delta[0]);
+			}
+		}
+		switch (valueType) {
+			case RANGE -> {changeRange(event, providedRange, mode);}
+			case PRIMARY -> {changePrimary(event, providedEffect, mode);}
+			case SECONDARY -> {changeSecondary(event, providedEffect, mode);}
+		}
+	}
+
+	private void changeRange(Event event, double providedRange, ChangeMode mode) {
+		switch (mode) {
+			case ADD -> {
+				for (Block block : getExpr().getArray(event)) {
+					Beacon beacon = (Beacon) block.getState();
+					beacon.setEffectRange(beacon.getEffectRange() + providedRange);
+					beacon.update(true);
+				}
+			}
+			case REMOVE -> {
+				for (Block block : getExpr().getArray(event)) {
+					Beacon beacon = (Beacon) block.getState();
+					beacon.setEffectRange(Math.max(beacon.getEffectRange() - providedRange, 0));
+					beacon.update(true);
+				}
+			}
+			case DELETE, RESET -> {
+				for (Block block : getExpr().getArray(event)) {
+					Beacon beacon = (Beacon) block.getState();
+					beacon.resetEffectRange();
```

Comment:
see https://discord.com/channels/135877399391764480/836220422223036467/1287060851848904785 for how to reduce duplicate code